### PR TITLE
Fallback to position-based progress and seeking if current time is unavailable

### PR DESCRIPTION
**Pull request by theli-ua**

### CODE REVIEW

File: src/libtomahawk/audio/AudioOutput.cpp [hunk 1]
```diff
@@ -37,7 +37,7 @@
 #include <vlc/libvlc_events.h>
 #include <vlc/libvlc_version.h>
 
-static const int ABOUT_TO_FINISH_TIME = 2000;
+static const float ABOUT_TO_FINISH_POS = 0.95f;
```
Comment (muesli) [inline src/libtomahawk/audio/AudioOutput.cpp]: This seems dangerous. Imagine listening to an hour long track - the "about to finish" would set in 3 minutes before the end of the song.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 PROJECT( tomahawk )
 CMAKE_MINIMUM_REQUIRED( VERSION 2.8.6 )
+set(CMAKE_EXPORT_COMPILE_COMMANDS "ON")
```
Comment (muesli) [inline CMakeLists.txt]: What is this required for?

File: src/libtomahawk/audio/AudioEngine.cpp [hunk 3]
```diff
@@ -401,10 +401,6 @@ AudioEngine::canSeek()
 {
     Q_D( AudioEngine );
 
-    if ( !d->audioOutput->isSeekable() ) {
```
Comment (muesli) [inline src/libtomahawk/audio/AudioEngine.cpp]: What happens when a stream really can't be seeked?

File: src/libtomahawk/audio/AudioOutput.cpp [hunk 4]
```diff
@@ -53,8 +53,8 @@ AudioOutput::AudioOutput( QObject* parent )
     : QObject( parent )
     , m_currentState( Stopped )
     , m_currentStream( nullptr )
-    , m_seekable( true )
     , m_muted( false )
+    , m_seekable( true )
```
Comment (muesli) [inline src/libtomahawk/audio/AudioOutput.cpp]: Doesn't this trigger a compiler warning about initialization order?

File: src/libtomahawk/audio/AudioOutput.cpp [hunk 5]
```diff
@@ -331,14 +346,22 @@ AudioOutput::setCurrentTime( qint64 time )
         total = AudioEngine::instance()->currentTrackTotalTime();
     }
 
-    if ( time < total - ABOUT_TO_FINISH_TIME )
-    {
-        m_aboutToFinish = false;
+    if(time <= 0) {
```
Comment (muesli) [inline src/libtomahawk/audio/AudioOutput.cpp]: Please try to stick to our coding style:

if ( time <= 0 )
{
...
}

File: src/libtomahawk/audio/AudioOutput.cpp [hunk 6]
```diff
@@ -331,14 +346,22 @@ AudioOutput::setCurrentTime( qint64 time )
...
+        m_seekable = false;
+    } else {
+        m_seekable = true;
     }
-    if ( !m_aboutToFinish && total > 0 && time >= total - ABOUT_TO_FINISH_TIME )
-    {
-        m_aboutToFinish = true;
-        emit aboutToFinish();
+
+    if(!m_havePosition) {
```
Comment (muesli) [inline src/libtomahawk/audio/AudioOutput.cpp]: Coding style.

File: src/libtomahawk/audio/AudioOutput.cpp [hunk 7]
```diff
@@ -428,18 +441,26 @@ AudioOutput::seek( qint64 milliseconds )
             return;
     }
 
-    //    tDebug() << Q_FUNC_INFO << "AudioOutput:: seeking" << milliseconds << "msec";
+    if( m_seekable ) {
```
Comment (muesli) [inline src/libtomahawk/audio/AudioOutput.cpp]: Coding style.

File: src/libtomahawk/audio/AudioEngine.cpp (hunk 3 above)
Comment (theli-ua) [inline src/libtomahawk/audio/AudioEngine.cpp]: With your current implementation isSeekable always returned true anyway because 
"Streams often can be seeked with imem even if reported as non seekable by vlc"

File: src/tomahawk/AudioControls.cpp [hunk 8]
```diff
@@ -469,7 +471,8 @@ AudioControls::onPlaybackSeeked( qint64 msec )
 {
     tDebug( LOGEXTRA ) << Q_FUNC_INFO;
     m_seeked = true;
-    onPlaybackTimer( msec );
+    if(m_haveTiming)
```
Comment (muesli) [inline src/tomahawk/AudioControls.cpp]: Coding style.

File: src/tomahawk/AudioControls.cpp [hunk 9]
```diff
@@ -517,6 +520,19 @@ AudioControls::onPlaybackStopped()
...
+    if (!m_haveTiming)
+    {
+        qint64 duration = AudioEngine::instance()->currentTrackTotalTime();
+        ui->seekSlider->blockSignals( true );
+        ui->seekSlider->setSliderPosition( position * duration );
+        ui->seekSlider->blockSignals( false );
+        m_sliderTimeLine.stop();
+    }
+}
+
```
Comment (muesli) [inline src/tomahawk/AudioControls.cpp]: Coding style: two new-lines methods methods.

File: CMakeLists.txt (hunk 2 above)
Comment (theli-ua) [inline CMakeLists.txt]: This is first commit in this pr. This can be used with things like https://github.com/Valloric/YouCompleteMe , which is a completion/etc plugin for Vim. 
See http://www.reddit.com/r/vim/comments/2aaoyr/youcompleteme_with_cmake_build_system/

File: src/libtomahawk/audio/AudioOutput.cpp (hunk 1 above)
Comment (theli-ua) [inline src/libtomahawk/audio/AudioOutput.cpp]: Not exactly sure why you use that "abouttofinish" thing anyway. If anything, even if track ended abruptly I personally would prefer it to go to the next song anyway.
Also, what would be dangerous about it being set 3 minutes before end?
And with current 2000 threshold I often get it not set for gmusic tracks.

File: src/libtomahawk/audio/AudioOutput.cpp (hunk 4 above)
Comment (theli-ua) [inline src/libtomahawk/audio/AudioOutput.cpp]: I don't think it does but I'll recheck

File: src/libtomahawk/audio/AudioOutput.cpp (hunk 1 above)
Comment (muesli) [inline src/libtomahawk/audio/AudioOutput.cpp]: I'm happy to get rid of it, if we can. It used to be important, but that stems from old phonon-vlc times.

File: src/libtomahawk/audio/AudioOutput.cpp [hunk 10]
```diff
@@ -331,14 +346,24 @@ AudioOutput::setCurrentTime( qint64 time )
         total = AudioEngine::instance()->currentTrackTotalTime();
     }
 
-    if ( time < total - ABOUT_TO_FINISH_TIME )
+    if( time <= 0 )
```
Comment (muesli) [inline src/libtomahawk/audio/AudioOutput.cpp]: Style: space missing after "if".

File: src/libtomahawk/audio/AudioOutput.cpp [hunk 11]
```diff
@@ -331,14 +346,24 @@ AudioOutput::setCurrentTime( qint64 time )
...
     {
-        m_aboutToFinish = false;
+        m_seekable = false;
+    } else {
+        m_seekable = true;
     }
-    if ( !m_aboutToFinish && total > 0 && time >= total - ABOUT_TO_FINISH_TIME )
+
+    if( !m_havePosition )
```
Comment (muesli) [inline src/libtomahawk/audio/AudioOutput.cpp]: Style: space missing after "if".

File: src/libtomahawk/audio/AudioOutput.cpp [hunk 12]
```diff
@@ -428,18 +443,29 @@ AudioOutput::seek( qint64 milliseconds )
             return;
     }
 
-    //    tDebug() << Q_FUNC_INFO << "AudioOutput:: seeking" << milliseconds << "msec";
+    if( m_seekable )
```
Comment (muesli) [inline src/libtomahawk/audio/AudioOutput.cpp]: Style: space missing after "if".

File: src/tomahawk/AudioControls.cpp [hunk 13]
```diff
@@ -469,7 +471,10 @@ AudioControls::onPlaybackSeeked( qint64 msec )
 {
     tDebug( LOGEXTRA ) << Q_FUNC_INFO;
     m_seeked = true;
-    onPlaybackTimer( msec );
+    if( m_haveTiming )
```
Comment (muesli) [inline src/tomahawk/AudioControls.cpp]: Style: space missing after "if".

File: CMakeLists.txt (hunk 2 above)
Comment (muesli) [inline CMakeLists.txt]: Ok, let's move that to a separate pull request, please.
